Print warning messages before cleanup runs.

fletcherm opened the pull request:
Note: this code was written largely by ChatGPT 4, so it may look a little different than the surrounding code. It was, of course, still inspected by a human. :)

What appears to be working:
* A warning message is sent 5 minutes before cleanup.

What appears to be missing:
* Even though the code is setup to send more messages every minute, we're only seeing that initial 5 minute warning message. This needs further investigation.

Next steps:
* Ensure a warning message is sent every minute before cleanup.
* (Possibly) adding some tests for this. This file doesn't already have a test suite so not clear to me if the juice is worth the squeeze.
* Compare the color of the cleanup warning text to MBBS proper. Right now we're using the same red color as the final cleanup message, but not sure if this is correct.
* Compare the literal text of the final cleanup message. My spidey sense is that does not match MBBS proper.

fletcherm commented:
@enusbaum @paladine This is a draft PR and not ready to merge. But I'd appreciate if you can take a look and at least see if you can identify when I'm only seeing the first warning message and not subsequent messages.

enusbaum commented on MBBSEmu/HostProcess/MbbsHost.cs:
```diff
@@ -212,7 +228,8 @@ public void Start(List<ModuleConfiguration> moduleConfigurations)
             _isRunning = false;
-            _timer.Dispose();
+            _cleanupTimer.Dispose();
+            _cleanupWarningTimer?.Dispose();
```
I think we should add the null check to both. While they're assigned in the constructor used most everywhere, if we decide to overload the constructor in the future (Unit Testing, etc.) we'll want to make sure this is gracefully handled.

Good catch!

fletcherm replied:
Should be all set now.

fletcherm commented:
![cleanup-warnings](https://github.com/mbbsemu/MBBSEmu/assets/147532/a7a035be-e2fd-4212-a7e1-39205332a90c)

Attached is a screenshot of what the cleanup warnings look like on WG.

fletcherm commented:
@enusbaum @paladine Alright, would love if you guys can take another look at this.

Only significant change since last round of reviews is that I changed how the cleanup time and grace period work, as we discussed in discord. (grace period comes _after_ cleanup time, for whatever reason) This should better match WG.

My bigger ask is that someone test & confirm the behavior that cleanup and the cleanup warning run more than once, and not just after the initial setup. Normally that would be once a day, of course, but feel free to tweak values to make it happen more frequently.

I _did_ do some experiments to test this myself, and things look good... but I had to hack enough lines of code to make cleanup more frequent that I kinda worry I changed something else incidentally and tainted the experiment. That is why I'd really appreciate another person testing it as well.

enusbaum commented on MBBSEmu/HostProcess/MbbsHost.cs:
```diff
@@ -1287,16 +1313,63 @@ private void DisableModule(string moduleId)
+            var dueTime = NowUntil(initialWarningTime);
+            var period = TimeSpan.FromMinutes(1);
+
+            Logger.Debug($"_cleanupTime Hours: {_cleanupTime.Hours}");
```
We can probably remove these. While they're helpful developing this feature, I don't see us needing these printed every startup in debug.

enusbaum commented on MBBSEmu/HostProcess/MbbsHost.cs:
```diff
@@ -1287,16 +1313,63 @@ private void DisableModule(string moduleId)
+                {
+                    var channel = c.Value.Channel;
+
+                    Logger.Error($"Sending cleanup warning to channel {channel}: {_cleanupWarningMinutesRemaining} {minuteText} until shutdown.");
```
This can be changed to DEBUG or INFO

enusbaum commented on MBBSEmu/HostProcess/MbbsHost.cs:
```diff
@@ -1287,16 +1313,63 @@ private void DisableModule(string moduleId)
+                    _channelDictionary[channel].SendToClient($"|RESET|\r\n|B||MAGENTA|Sorry to interrupt here, but the server will be shutting down in {_cleanupWarningMinutesRemaining} {minuteText} for the nightly \"auto-cleanup\" process. Please finish up and log off... thank you!|RESET|\r\n".EncodeToANSIArray());
+                }
+                _cleanupWarningMinutesRemaining--;
+                Logger.Debug($"finished SendCleanupWarning");
```
Can be removed

enusbaum commented on MBBSEmu/HostProcess/MbbsHost.cs:
```diff
@@ -96,14 +96,36 @@ public class MbbsHost : IMbbsHost, IDisposable
+        /// <summary>
+        ///     Amount of time to give as a grace period to logoff after nightly cleanup.
+        /// </summary>
+        private const int _cleanupGracePeriodMinutes = 10;
```
Following our naming for constants, this should be `CleanupGracePeriodMinutes`

enusbaum commented on MBBSEmu/HostProcess/MbbsHost.cs:
```diff
@@ -1287,16 +1313,63 @@ private void DisableModule(string moduleId)
+
+                var minuteText = (_cleanupWarningMinutesRemaining > 1) ? "minutes" : "minute";
+
+                foreach (var c in _channelDictionary)
```
You can combine the channel selection on the next line and the `foreach` definition into one line using a LINQ expression for the values being iterated over: 

`foreach (var channel in _channelDictionary.Select(c => c.Value.Channel))`
